File: kattis/union-find.py

N, Q = map(int, input().split())

par = []
for i in range(N):
    par.append(i)

rank = [0] * N

def get_root(a):
    parent = par[a]
    if parent == a:
        return a
    else:
        par[a] = get_root(parent)
        return par[a]

def join_set(a, b):
    ra = get_root(a)
    rb = get_root(b)
    if ra == rb: # sets are already joined
        return

    # keep the set that's tallest
    if rank[rb] > rank[ra]:
        join_set(b,a)
    else:
        par[rb] = ra

    # if both sets have the same rank, then the one that becomes root will have its rank incremented by 1
    if (rank[ra] == rank[rb]):
        rank[ra] += 1


for i in range(Q):
    line = input()
    op, a, b = line.split()
    a = int(a)
    b = int(b)

    if op == "?":
        if (get_root(a) == get_root(b)):
            print("yes")
        else:
            print("no")

    elif op == "=":
        join_set(a, b)

File: kattis/union-find.cpp
#include <iostream>

using namespace std;

int par[1000000];
int rnk[1000000];

int find_root(int a); // find root of a node
void check_elems(int a, int b); // check if two nodes are in the same set
void join(int a, int b); // join the sets to which a and b belong

int main(){

    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

    int N, Q; cin >> N >> Q;
    for (int i = 0; i < N; i++){
        par[i] = i;
        rnk[i] = 0;
    }

    // Read in operations
    char op; int a, b;
    for (int i = 0; i < Q; i++){
        cin >> op >> a >> b;
        if (op == '?')
            check_elems(a, b);
        else if (op == '=')
            join(a, b);
    }
}

// Recursively find representative ("root") node of a set, given an element that belongs to that set
int find_root(int a){
    if (a == par[a])
        return a;
    else{
        return par[a] = find_root(par[a]);
        //return par[a];
    }
}

// check if two elements are in the same set
void check_elems(int a, int b){
    if (find_root(a) == find_root(b))
        cout << "yes\n";
    else
        cout << "no\n";
}

void join(int a, int b){
    int ra = find_root(a);
    int rb = find_root(b);
    // if the elemnts are already in the same set, do nothing
    if (ra == rb)
        return;
    // we assume a has higher rank
    if (rnk[ra] < rnk[rb])
        join(b, a);
    else{ // rank[a] >= rank[b] -> we join to a
        par[rb] = ra;
        if (rnk[ra] == rnk[rb]) // if rank[ra] == rank[rb], the height of set ra increases by 1
            rnk[ra] += 1;
        return;
    }
}

